File: leetcode/javascript/array/TwoPointers/isSubsequence(easy)/isSubsequence.js
/*
Учитывая две строки s и t, верните true, если s является подпоследовательностью t, или false в противном случае.

Подпоследовательность строки — это последовательность символов, которую можно получить путем удаления некоторых
(или ни одного) символов из исходной строки, сохраняя при этом относительный порядок оставшихся символов.
Например, «ace» является подпоследовательностью «abcde», а «aec» — нет.
 */

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isSubsequence = function(s, t) {
    let left = 0, right = 0;
    while (left < s.length && right < t.length) {
        if (s[left] === t[right]) left++;
        right++;
    }
    return left === s.length;
};

console.log(isSubsequence("ace", "abcde")); // true
console.log(isSubsequence("aec", "abcde")); // false

File: leetcode/go/array/TwoPointers/isSubsequence(easy)/isSubsequence.go
package main

import "fmt"

/*
Учитывая две строки s и t, верните true, если s является подпоследовательностью t, или false в противном случае.

Подпоследовательность строки — это последовательность символов, которую можно получить путем удаления некоторых
(или ни одного) символов из исходной строки, сохраняя при этом относительный порядок оставшихся символов.
Например, «ace» является подпоследовательностью «abcde», а «aec» — нет.
*/

func main() {
	fmt.Println(isSubsequence("ace", "abcde")) // true
	fmt.Println(isSubsequence("aec", "abcde")) // false
}

func isSubsequence(s, t string) bool {
	left, right := 0, 0

	for left < len(s) && right < len(t) {
		if s[left] == t[right] {
			left++
		}
		right++
	}

	return left == len(s)
}
